### PR TITLE
Add hotspot vouchers

### PR DESCRIPTION
**Pull request by ufozone**

### CODE REVIEW

**Review by Kane610 — CHANGES_REQUESTED** [review]
Thanks for the PR

Please add tests, please add to the gateway class, recommended is to also setup the pre-commit hook with setup.sh and you will catch a lot of small things

**Comment by ufozone** [comment]
@Kane610, I am unfortunately very new to Phyton / Home Assistant development and have no idea how to do add tests, add to the gateway class and so on.

@bmos has made some changes with regard to code hygiene. Thanks for that!

**Comment by Kane610** [comment]
> @Kane610, I am unfortunately very new to Phyton / Home Assistant development and have no idea how to do add tests, add to the gateway class and so on.
> 
> @bmos has made some changes with regard to code hygiene. Thanks for that!

Here is a PR adding a new ApiHandler based class to the controller, adds api requests and tests, that should help you to get going

https://github.com/Kane610/aiounifi/commit/68ee89d27a154f609b95fc5c7bab126b52cda0e9

Don't hesitate to ask questions if you get stuck

**Comment by ufozone** [comment]
I have now built something for the PR from the linked PR. I didn't understand anything about the tests. I don't know if that's enough or how to test it.

I have no idea what I'm doing

**Comment by Kane610** [comment]
Skimmed it through, will review during the weekend

File: aiounifi/models/voucher.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    @property
+    def status(self) -> str:
+        """Status."""
+        return self.raw.get("status", "")
+
+    @property
+    def status_expires(self) -> timedelta | None:
+        """Status expires."""
+        if self.raw.get("status_expires", 0.0) > 0:
+            return timedelta(seconds=self.raw.get("status_expires", 0.0))
```
Comment (Kane610) [inline aiounifi/models/voucher.py:215]:
```suggestion
        if (status_expiry := self.raw.get("status_expires", 0.0)) > 0:
            return timedelta(seconds=status_expiry)
```

**Review by Kane610 — CHANGES_REQUESTED** [review]
Really nice! Just one small thing to improve from review and you need to fix the failing validation

**Comment by ufozone** [comment]
@Kane610 , thank you very much. The suggested change has been implemented.

File: aiounifi/models/voucher.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    qos_rate_max_up: int
+    qos_rate_max_down: int
+    used: int
+    create_time: float
+    start_time: float
+    end_time: float
+    for_hotspot: bool
+    admin_name: str
+    status: str
+    status_expires: float
```
Comment (Kane610) [inline aiounifi/models/voucher.py:30]: You have a lot of attributes which seems to be optional in `class Voucher(ApiItem):` if they are expected to be omitted in certain cases please mark those with NotRequired

File: aiounifi/models/voucher.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+    @property
+    def id(self) -> str:
+        """ID of voucher."""
+        return self.raw["_id"]
+
+    @property
+    def site_id(self) -> str:
+        """Site ID."""
+        return self.raw["_id"]
```
Comment (Kane610) [inline aiounifi/models/voucher.py:130]: Is this superfluous?

File: aiounifi/models/voucher.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+    @property
+    def note(self) -> str:
+        """Note."""
```
Comment (Kane610) [inline aiounifi/models/voucher.py:134]: Please use more than one word in the doc strings

File: aiounifi/models/voucher.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        return self.raw.get("note") or ""
+
+    @property
+    def code(self) -> str:
+        """Code."""
+        if len(c := self.raw.get("code", "")) > 5:
+            return f"{c[:5]}-{c[5:]}"
+        return c
```
Comment (Kane610) [inline aiounifi/models/voucher.py:142]: What is happening here, please document

File: aiounifi/models/voucher.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+    @property
+    def qos_usage_quota(self) -> int:
+        """Quantity of bytes allowed in MB."""
+        return int(self.raw.get("qos_usage_quota", 0))
+
+    @property
+    def qos_rate_max_up(self) -> int:
+        """Up speed allowed in kbps."""
+        return int(self.raw.get("qos_rate_max_up", 0))
```
Comment (Kane610) [inline aiounifi/models/voucher.py:167]: This shouldn't need int()

File: aiounifi/models/voucher.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+    @property
+    def qos_overwrite(self) -> bool:
+        """Used count."""
+        return self.raw.get("qos_overwrite", False)
+
+    @property
+    def qos_usage_quota(self) -> int:
+        """Quantity of bytes allowed in MB."""
+        return int(self.raw.get("qos_usage_quota", 0))
```
Comment (Kane610) [inline aiounifi/models/voucher.py:162]: If typing is correct you shouldn't need to use int in these places , should it really be str in TypedVoucher?

File: aiounifi/models/voucher.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+    @property
+    def qos_rate_max_up(self) -> int:
+        """Up speed allowed in kbps."""
+        return int(self.raw.get("qos_rate_max_up", 0))
+
+    @property
+    def qos_rate_max_down(self) -> int:
+        """Down speed allowed in kbps."""
+        return int(self.raw.get("qos_rate_max_down", 0))
```
Comment (Kane610) [inline aiounifi/models/voucher.py:172]: This shouldn't need int()

**Review by Kane610 — CHANGES_REQUESTED** [review]
Nearly there! Good job! 

Some small areas of test coverage should be improved, I think there is no coverage of deletion path for example

File: aiounifi/models/voucher.py (hunk 3 above)
Comment (ufozone) [reply aiounifi/models/voucher.py:130]: No

File: aiounifi/models/voucher.py (hunk 2 above)
Comment (ufozone) [reply aiounifi/models/voucher.py]: Good advice. Done.

File: aiounifi/models/voucher.py (hunk 4 above)
Comment (ufozone) [reply aiounifi/models/voucher.py:134]: Done

File: aiounifi/models/voucher.py (hunk 7 above)
Comment (ufozone) [reply aiounifi/models/voucher.py:162]: qos_usage_quota comes from the API unnecessarily as a string

File: aiounifi/models/voucher.py (hunk 6 above)
Comment (ufozone) [reply aiounifi/models/voucher.py:167]: Ok. I wanted to produce strict typing. Doesn't seem to be necessary?

File: aiounifi/models/voucher.py (hunk 8 above)
Comment (ufozone) [reply aiounifi/models/voucher.py:172]: Done

**Comment by ufozone** [comment]
@Kane610 , I have made adjustments.

I work on the tests blindfolded. If other things are missing here, I need support.

File: tests/test_vouchers.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    await vouchers.update()
+
+    assert unifi_called_with("get", "/api/s/default/stat/voucher")
+    assert len(vouchers.values()) == 0
+
+
+@pytest.mark.parametrize("voucher_payload", [VOUCHERS])
+@pytest.mark.usefixtures("_mock_endpoints")
+async def test_vouchers(
+    mock_aioresponse, unifi_controller, _mock_endpoints, unifi_called_with
```
Comment (Kane610) [inline tests/test_vouchers.py:68]:
```suggestion
    mock_aioresponse, unifi_controller, unifi_called_with
```

File: tests/test_vouchers.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        "/api/s/default/cmd/hotspot",
+        json={
+            "cmd": "delete-voucher",
+            "_id": "657e370a4543a555901865c7",
+        },
+    )
+
+
+@pytest.mark.usefixtures("_mock_endpoints")
+async def test_no_vouchers(unifi_controller, _mock_endpoints, unifi_called_with):
```
Comment (Kane610) [inline tests/test_vouchers.py:56]:
```suggestion
async def test_no_vouchers(unifi_controller, unifi_called_with):
```

**Review by Kane610 — CHANGES_REQUESTED** [review]
In conftest.py you're gonna need to add voucher data in a couple of places

_endpoint_fixture should have a input like
`vouchers_payload: list[dict[str, Any]]`
and a mock_get_request like
```python
    mock_get_request(
        "/api/s/default/stat/voucher",
        "/proxy/network/api/s/default/stat/voucher",
        vouchers_payload,
    )
```


and you also need a separate fixture 

```python
@pytest.fixture(name="vouchers_payload")
def vouchers_data_fixture() -> list[dict[str, Any]]:
    """Vouchers data."""
    return []
```

**Comment by Kane610** [comment]
> @Kane610 , I have made adjustments.
> 
> I work on the tests blindfolded. If other things are missing here, I need support.

You're learning! :D

**Comment by ufozone** [comment]
These tests are still killing me. With small steps we are getting closer to the goal ... hopefully.

@Kane610 , take a look, I've done _things_ again.

**Comment by Kane610** [comment]
If you can it's better if you can setup the environment locally so you can run the different tests, on linux it should only be to run the ./setup.sh  and then source venv/bin/activate and you're running it

File: tests/test_vouchers.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+    assert unifi_called_with("get", "/api/s/default/stat/voucher")
+    assert len(vouchers.values()) == 0
+
+
+@pytest.mark.parametrize("voucher_payload", [VOUCHERS])
+@pytest.mark.usefixtures("_mock_endpoints")
+async def test_vouchers(
+    mock_aioresponse, unifi_controller, unifi_called_with
+):
```
Comment (Kane610) [inline tests/test_vouchers.py:69]:
```suggestion
async def test_vouchers(mock_aioresponse, unifi_controller, unifi_called_with):
```

File: tests/conftest.py [hunk 12]
```diff
@@ -255,3 +261,9 @@ def traffic_rule_status_fixture() -> int:
 def wlan_data_fixture() -> list[dict[str, Any]]:
     """WLAN data."""
     return []
+
+
+@pytest.fixture(name="vouchers_payload")
+def vouchers_data_fixture() -> list[dict[str, Any]]:
```
Comment (Kane610) [inline tests/conftest.py:267]:
```suggestion
def voucher_data_fixture() -> list[dict[str, Any]]:
```
lets do singular on these ones

File: tests/conftest.py [hunk 13]
```diff
@@ -255,3 +261,9 @@ def traffic_rule_status_fixture() -> int:
 def wlan_data_fixture() -> list[dict[str, Any]]:
     """WLAN data."""
     return []
+
+
+@pytest.fixture(name="vouchers_payload")
```
Comment (Kane610) [inline tests/conftest.py:266]:
```suggestion
@pytest.fixture(name="voucher_payload")
```

File: tests/conftest.py [hunk 14]
```diff
@@ -177,6 +178,11 @@ def mock_get_request(
         "/proxy/network/api/s/default/rest/wlanconf",
         wlan_payload,
     )
+    mock_get_request(
+        "/api/s/default/stat/voucher",
+        "/proxy/network/api/s/default/stat/voucher",
+        vouchers_payload,
```
Comment (Kane610) [inline tests/conftest.py:184]:
```suggestion
        voucher_payload,
```

File: tests/conftest.py [hunk 15]
```diff
@@ -103,6 +103,7 @@ def _endpoint_fixture(
     traffic_rule_payload: list[dict[str, Any]],
     traffic_rule_status: int,
     wlan_payload: list[dict[str, Any]],
+    vouchers_payload: list[dict[str, Any]],
```
Comment (Kane610) [inline tests/conftest.py:106]:
```suggestion
    voucher_payload: list[dict[str, Any]],
```

File: tests/test_controller.py [hunk 16]
```diff
@@ -284,6 +285,7 @@ async def test_controller(unifi_controller, unifi_called_with, new_ws_data_fn):
     assert len(unifi_controller.system_information.items()) == 0
     assert len(unifi_controller.traffic_routes.items()) == 0
     assert len(unifi_controller.traffic_rules.items()) == 0
+    assert len(unifi_controller.voucher.items()) == 0
```
Comment (Kane610) [inline tests/test_controller.py:288]:
```suggestion
    assert len(unifi_controller.vouchers.items()) == 0
```

File: tests/test_vouchers.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    """Test that different types of ports work."""
+    vouchers = unifi_controller.vouchers
+    await vouchers.update()
+    assert len(vouchers.values()) == 2
+
+    voucher = vouchers["657e370a4543a555901865c7"]
+    assert voucher.id == "657e370a4543a555901865c7"
+    assert voucher.site_id == "5a32aa4ee4b0412345678910"
```
Comment (Kane610) [inline tests/test_vouchers.py:77]:
```suggestion
    assert voucher.site_id == "657e370a4543a555901865c7"
```

File: aiounifi/models/voucher.py [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+    @property
+    def note(self) -> str:
+        """Note describing voucher."""
+        return self.raw.get("note") or ""
+
+    @property
+    def code(self) -> str:
+        """Code in known format 00000-00000."""
+        if len(c := self.raw.get("code", "")) > 5:
```
Comment (Kane610) [inline aiounifi/models/voucher.py:140]:
```suggestion
        if len(c := str(self.raw.get("code", ""))) > 5:
```
In your voucher data code can be an int

File: aiounifi/models/voucher.py [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+from .api import ApiItem, ApiRequest
+
+
+class TypedVoucher(TypedDict):
+    """Voucher type definition."""
+
+    _id: str
+    site_id: str
+    note: str
+    code: str
```
Comment (Kane610) [inline aiounifi/models/voucher.py:16]:
```suggestion
    code: int | str
```

File: tests/fixtures.py [hunk 20]
```diff
@@ -4110,6 +4110,39 @@
...
+        "for_hotspot": "false",
+        "note": "auto-generated",
+        "qos_overwrite": "false",
+        "quota": 0,
+        "site_id": "5a32aa4ee4b0412345678910",
+        "status": "USED_MULTIPLE",
+        "status_expires": 244679302,
+        "used": 2,
+    },
+]
```
Comment (Kane610) [inline tests/fixtures.py:4144]:
```suggestion
VOUCHERS = [
    {
        "_id": "657e370a4543a555901865c7",
        "admin_name": "Admin",
        "code": 7470075124,
        "create_time": 1702770442,
        "duration": 480,
        "for_hotspot": False,
        "note": "auto-generated",
        "qos_overwrite": False,
        "quota": 1,
        "site_id": "5a32aa4ee4b0412345678910",
        "status": "VALID_ONE",
        "status_expires": 0,
        "used": 0,
    },
    {
        "_id": "61facea3873fdb075ce28d71",
        "admin_name": "Admin",
        "code": 4470374222,
        "create_time": 1643826851,
        "duration": 5184000,
        "for_hotspot": False,
        "note": "auto-generated",
        "qos_overwrite": False,
        "quota": 0,
        "site_id": "5a32aa4ee4b0412345678910",
        "status": "USED_MULTIPLE",
        "status_expires": 244679302,
        "used": 2,
    },
]
```
Please double check with real data, json data when read out typically shows true false without capital letters but in python they should be changed to True/False

File: tests/test_vouchers.py [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+    assert voucher.qos_rate_max_up == 0
+    assert voucher.qos_rate_max_down == 0
+    assert voucher.used == 0
+    assert voucher.create_time == datetime.fromtimestamp(1702770442)
+    assert voucher.start_time is None
+    assert voucher.end_time is None
+    assert voucher.for_hotspot is False
+    assert voucher.admin_name == "Admin"
+    assert voucher.status == "VALID_ONE"
+    assert voucher.status_expires == 0
```
Comment (Kane610) [inline tests/test_vouchers.py:93]:
```suggestion
    assert voucher.status_expires is None
```

**Review by Kane610 — CHANGES_REQUESTED** [review]
Some more places breaking in the tests

**Comment by ufozone** [comment]
Wow, thanks for your great support!

File: aiounifi/models/voucher.py [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+    def start_time(self) -> datetime | None:
+        """Start datetime."""
+        if "start_time" in self.raw:
+            return datetime.fromtimestamp(self.raw["start_time"])
+
+    @property
+    def end_time(self) -> datetime | None:
+        """End datetime."""
+        if "end_time" in self.raw:
+            return datetime.fromtimestamp(self.raw["end_time"])
```
Comment (Kane610) [inline aiounifi/models/voucher.py:195]:
```suggestion
    @property
    def start_time(self) -> datetime | None:
        """Start datetime."""
        if "start_time" in self.raw:
            return datetime.fromtimestamp(self.raw["start_time"])
        return None

    @property
    def end_time(self) -> datetime | None:
        """End datetime."""
        if "end_time" in self.raw:
            return datetime.fromtimestamp(self.raw["end_time"])
        return None
```

File: aiounifi/models/voucher.py [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+    qos_rate_max_up: NotRequired[int]
+    qos_rate_max_down: NotRequired[int]
+    used: int
+    create_time: float
+    start_time: NotRequired[float]
+    end_time: NotRequired[float]
+    for_hotspot: bool
+    admin_name: str
+    status: str
+    status_expires: float
```
Comment (Kane610) [inline aiounifi/models/voucher.py:30]: The amount of raw.gets in the class properties are more than the amount of NotRequired here

**Review by Kane610 — CHANGES_REQUESTED** [review]

File: aiounifi/models/voucher.py (hunk 23 above)
Comment (ufozone) [reply aiounifi/models/voucher.py:30]: Sounds legit. I had tested much, but does not know if I have found all required and optional fields.
I will change it in one hour

Comment (ufozone) [reply aiounifi/models/voucher.py]: Now everything should fit. I had an issue with the "code" attribute in the fixture. The "code" only returns as a string.

**Comment by ufozone** [comment]
✅ Voucher type definition properties and class methods adapted.
✅ Fix test fixture for voucher response.
✅ Minor bugs fixed and documentation improved

🤞Cross your fingers that the checks go through and we are ready for take off.

**Comment by Kane610** [comment]
Test fail, looks like you missed changing an id

File: tests/test_vouchers.py [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+@pytest.mark.usefixtures("_mock_endpoints")
+async def test_vouchers(mock_aioresponse, unifi_controller, unifi_called_with):
+    """Test that different types of ports work."""
+    vouchers = unifi_controller.vouchers
+    await vouchers.update()
+    assert len(vouchers.values()) == 2
+
+    voucher = vouchers["657e370a4543a555901865c7"]
+    assert voucher.id == "657e370a4543a555901865c7"
+    assert voucher.site_id == "657e370a4543a555901865c7"
```
Comment (Kane610) [inline tests/test_vouchers.py:75]:
```suggestion
    assert voucher.site_id == "5a32aa4ee4b0412345678910"
```

**Review by Kane610 — CHANGES_REQUESTED** [review]
You have some coverage left to do, don't fret, you have done most of the heavy lifting already

```
aiounifi/interfaces/vouchers.py   14    2    86%   25, 40
aiounifi/models/voucher.py          112   8   93%   81, 83, 85, 87, 143, 189, 196, 218
```

**Comment by ufozone** [comment]
I first had to read up on what this coverage is. I'll give it a try.

**Review by ufozone — COMMENTED** [review]
🤞

**Review by ufozone — COMMENTED** [review]
So I'm done. Mentally and with the code.

**Review by Kane610 — CHANGES_REQUESTED** [review]
With these changes the tests should pass and coverage be 100%

diff --git a/tests/test_vouchers.py b/tests/test_vouchers.py
index 644ead6..ee81ba9 100644
--- a/tests/test_vouchers.py
+++ b/tests/test_vouchers.py
@@ -41,9 +41,9 @@ async def test_voucher_create(mock_aioresponse, unifi_controller, unifi_called_w
             "quota": 0,
             "expire_number": 3600,
             "expire_unit": 1,
-            "usage_quota": 1000,
-            "rate_max_up": 5000,
-            "rate_max_down": 2000,
+            "bytes": 1000,
+            "up": 5000,
+            "down": 2000,
             "note": "Unit Testing",
         },
     )
@@ -108,7 +108,7 @@ async def test_vouchers(mock_aioresponse, unifi_controller, unifi_called_with):
     voucher = vouchers["61facea3873fdb075ce28d71"]
     assert voucher.id == "61facea3873fdb075ce28d71"
     assert voucher.site_id == "5a32aa4ee4b0412345678910"
-    assert voucher.note is None
+    assert voucher.note == ""
     assert voucher.code == "44703"
     assert voucher.quota == 1
     assert voucher.duration == timedelta(minutes=480)
@@ -138,9 +138,12 @@ async def test_vouchers(mock_aioresponse, unifi_controller, unifi_called_with):
         json={
             "cmd": "create-voucher",
             "n": 1,
-            "quota": 0,
-            "expire_number": 3600,
+            "quota": 1,
+            "expire_number": 480,
             "expire_unit": 1,
+            "bytes": 1000,
+            "up": 2000,
+            "down": 5000,
         },
     )
 
@@ -150,6 +153,6 @@ async def test_vouchers(mock_aioresponse, unifi_controller, unifi_called_with):
         "/api/s/default/cmd/hotspot",
         json={
             "cmd": "delete-voucher",
-            "_id": "657e370a4543a555901865c7",
+            "_id": "61facea3873fdb075ce28d71",
         },
     )

**Comment by Kane610** [comment]
You should change to do it locally

**Comment by ufozone** [comment]
woohoo 100% coverage.


> You should change to do it locally

Oh yeah, I didn't realize how deep I was falling down the rabbit hole last year. I just wanted to write a little integration. Now there are three and we're in this PR here.

**Comment by Kane610** [comment]
I want ro go through it one last time. Jut I don't have time today and busy a few days so it would have to be on Monday

File: aiounifi/models/voucher.py [hunk 25]
```diff
@@ -0,0 +1,219 @@
...
+
+    @property
+    def note(self) -> str:
+        """Note describing voucher."""
+        return self.raw.get("note") or ""
+
+    @property
+    def code(self) -> str:
+        """Code in known format 00000-00000."""
+        if len(c := self.raw.get("code", "")) > 5:
```
Comment (Kane610) [inline aiounifi/models/voucher.py:140]: Why is it specifically 5 that you're comparing against? Couldn't it be that if the code isn't the length of 10 characters it should just be returned?

Comment (Kane610) [reply aiounifi/models/voucher.py:140]: If you could update the pydoc string with a bit more of these details then we can merge this work

**Comment by ufozone** [comment]
So last commit

File: aiounifi/models/voucher.py [hunk 26]
```diff
@@ -0,0 +1,227 @@
...
+        return self.raw["site_id"]
+
+    @property
+    def note(self) -> str:
+        """Note given by user to voucher."""
+        return self.raw.get("note") or ""
+
+    @property
+    def code(self) -> str:
+        """Code in known format 00000-00000."""
```
Comment (Kane610) [inline aiounifi/models/voucher.py:139]: Expand this pydoc with the details about the different expected types of code as you responded in my previous question

Comment (ufozone) [reply aiounifi/models/voucher.py]: That must be enough now

Comment (ufozone) [reply aiounifi/models/voucher.py]: I don't understand the problem. The code is easy to understand. If the returned code is longer than five characters, then separate the code after the fifth character. If the code is not five characters long, it cannot be separated there. Anyone who doesn't understand this should perhaps not look at source code in general....

**Comment by ufozone** [comment]
Last attempt. Merge it or close this PR. It is difficult for me to document self-explanatory code in more detail.

**Comment by Kane610** [comment]
> Last attempt. Merge it or close this PR. It is difficult for me to document self-explanatory code in more detail.

Maybe I'm using bad wording to express myself here and I apologize if that's the case. This should be read as me trying to motivate my issue with the current description in pydoc, what I have an understanding of what I'm missing in it based on your explanation and why this is important for me in the long term.

Look here. The comment says "Code in known format [00000-00000](tel:00000-00000)". If the code where to be self explanatory it would not do anything but making sure that there is a hyphen in the middle. Now the code does something else as well, which is not obvious to me either what that means. It should be obvious why these two paths exist. You had to play around with it yourself to figure out the five character limit.

I'm more or less asking you to rephrasing this "I experimented a bit and received vouchers with 5, 8 and regular 10 characters. When entering the code on the hotspot page, the hyphen must be placed after the fifth digit.". I don't think this is an unreasonable request worth getting defensive over.

I've previously accepted PRs with code that when I needed to change things had to spend a lot of time to understand to when refactoring things. If functionality is not clear then it will come back as a personal cost on me as I'm the one sitting on long term maintenance of the code base. That's why I want to make sure to make new mistakes and learn from old.

I hope I've explained myself enough and you have an understanding behind my motivations with this request.

/R

**Comment by ufozone** [comment]
In the Hotspot Manager UI, the regular 10-digit code is displayed with the hyphen "in the middle". However, the unofficial API delivers it without this hyphen.

Now it would be simple to play out the code in the model as it comes from the API.

However, the captive portal expects the hyphen to be entered. So I build in the code the hyphen "in the middle" - f"{c[:5]}-{c[5:]}". 

Through some nasty manipulation, I managed to get back a five and an eight-digit code. Putting a hyphen in there seems to be wrong. But maybe not. I don't know. Maybe it's just an edge case. There was no rocket science in the position of the hyphen.

That's why I've now simplified the code. If it cracks at some point, we'll work it out then.

**Comment by Kane610** [comment]
Test fails

FAILED tests/test_vouchers.py::test_vouchers[voucher_payload0] - AssertionError: assert '44703-' == '44703'
  
  - 44703
  + 44703-

Will the output from unifi be with or without hyphen?

File: aiounifi/interfaces/vouchers.py [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+    async def create(self, voucher: Voucher) -> TypedApiResponse:
+        """Create voucher on controller."""
+        return await self.controller.request(
+            VoucherCreateRequest.create(
+                number=1,
+                quota=voucher.quota,
+                expire_number=int(
+                    voucher.duration.total_seconds() / 60  # Get minutes.
+                ),
+                expire_unit=1,
```
Comment (Kane610) [inline aiounifi/interfaces/vouchers.py:29]: As the default value is 1 for expire unit would this be ok, or is there a specific reason this is 1
```suggestion
                expire_unit=voucher.expire_unit,
```

File: aiounifi/interfaces/vouchers.py [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+
+    obj_id_key = "_id"
+    item_cls = Voucher
+    api_request = VoucherListRequest.create()
+
+    async def create(self, voucher: Voucher) -> TypedApiResponse:
+        """Create voucher on controller."""
+        return await self.controller.request(
+            VoucherCreateRequest.create(
+                number=1,
```
Comment (Kane610) [inline aiounifi/interfaces/vouchers.py:24]: Number is a required item to create an instance of Voucher, is there a reason 1 is hardcoded?

**Review by Kane610 — COMMENTED** [review]
Ok, to merge once test pass, two questions for completeness

File: aiounifi/interfaces/vouchers.py (hunk 28 above)
Comment (ufozone) [reply aiounifi/interfaces/vouchers.py:24]: Number is not an attribute of Voucher. Number is the number of vouchers to be create in one request

**Comment by Kane610** [comment]
Right now the test expects "44703-"

File: aiounifi/interfaces/vouchers.py (hunk 27 above)
Comment (ufozone) [reply aiounifi/interfaces/vouchers.py:29]: exire_unit is not an attribute of Voucher. 1 = minute, because expire_number came from voucher.duration in Seconds / 60 = Minutes

The attributes for vouchers of the return values of the API are called differently than those for creating.
For example, when creating, you pass the number of vouchers to be created. But these are then created as separate vouchers.
And when creating, you specify the expiry time and in which unit you specify the expiry time. In the return, you always get the duration time in seconds. However, you cannot create a voucher in seconds ....

Comment (Kane610) [reply aiounifi/interfaces/vouchers.py:29]: Alright, yes indeed, thanks, as it has the default value assigned this can just be removed.
```suggestion
```

File: aiounifi/interfaces/vouchers.py (hunk 28 above)
Comment (Kane610) [reply aiounifi/interfaces/vouchers.py:24]: Alright, I would just add the default value  to the create request and remove this instead.
```suggestion
```

File: aiounifi/models/voucher.py [hunk 29]
```diff
@@ -0,0 +1,229 @@
...
+@dataclass
+class VoucherCreateRequest(ApiRequest):
+    """Request object for voucher create."""
+
+    @classmethod
+    def create(
+        cls,
+        number: int,
+        quota: int,
+        expire_number: int,
```
Comment (Kane610) [inline aiounifi/models/voucher.py:55]:
```suggestion
        expire_number: int,
        number: int = 1,
        quota: int = 0,
```

File: aiounifi/models/voucher.py [hunk 30]
```diff
@@ -0,0 +1,229 @@
...
+        :param note: description
+        """
+        data = {
+            "cmd": "create-voucher",
+            "n": number,
+            "quota": quota,
+            "expire_number": expire_number,
+            "expire_unit": expire_unit,
+        }
+        if usage_quota:
```
Comment (Kane610) [inline aiounifi/models/voucher.py:80]:
```suggestion
        if usage_quota is not None:
```
Unless 0 is not allowed then its ok

File: aiounifi/models/voucher.py [hunk 31]
```diff
@@ -0,0 +1,229 @@
...
+            data["bytes"] = usage_quota
+        if rate_max_up:
```
Comment (Kane610) [inline aiounifi/models/voucher.py:82]:
```suggestion
        if rate_max_up is not None:
```
Unless 0 is not allowed then its ok

File: aiounifi/models/voucher.py [hunk 32]
```diff
@@ -0,0 +1,229 @@
...
+            data["up"] = rate_max_up
+        if rate_max_down:
```
Comment (Kane610) [inline aiounifi/models/voucher.py:84]:
```suggestion
        if rate_max_down is not None:
```
Unless 0 is not allowed then its ok

**Review by Kane610 — APPROVED** [review]

File: aiounifi/models/voucher.py (hunk 30 above)
Comment (ufozone) [reply aiounifi/models/voucher.py:80]: 0 is not allowed.

File: aiounifi/models/voucher.py (hunk 31 above)
Comment (ufozone) [reply aiounifi/models/voucher.py:82]: 0 is not allowed.

File: aiounifi/models/voucher.py (hunk 32 above)
Comment (ufozone) [reply aiounifi/models/voucher.py:84]: 0 is not allowed.

**Comment by Kane610** [comment]
Suggested changes in https://github.com/Kane610/aiounifi/pull/554

File: aiounifi/models/voucher.py (hunk 29 above)
Comment (Kane610) [reply aiounifi/models/voucher.py:55]: Thanks
